File: gateway/evals/skill_router_trend_meta.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any


ZERO_SHA = "0000000000000000000000000000000000000000"


def _normalize_optional_text(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    normalized = value.strip()
    if not normalized or normalized == ZERO_SHA:
        return None
    return normalized


def _normalize_optional_bool(value: Any) -> bool | None:
    if isinstance(value, bool):
        return value
    if not isinstance(value, str):
        return None
    normalized = value.strip().lower()
    if normalized == "true":
        return True
    if normalized == "false":
        return False
    return None


def _load_report(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    with path.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)
    if isinstance(payload, dict):
        return payload
    return {}


def _extract_policy_hash(report: dict[str, Any]) -> str | None:
    policy = report.get("policy")
    if isinstance(policy, dict):
        hash_value = _normalize_optional_text(policy.get("hash"))
        if hash_value is not None:
            return hash_value
    return _normalize_optional_text(report.get("policy_hash"))


def build_skill_router_trend_meta(
    *,
    current_report: dict[str, Any],
    base_report: dict[str, Any],
    trend_mode: str,
    trend_reason: str,
    trend_required: Any,
    baseline_available: Any,
    base_sha: Any,
    current_policy_blob: Any,
    base_policy_blob: Any,
    policy_blob_match: Any,
) -> dict[str, Any]:
    current_policy_hash = _extract_policy_hash(current_report)
    base_policy_hash = _extract_policy_hash(base_report)

    policy_hash_match: bool | None = None
    if current_policy_hash is not None and base_policy_hash is not None:
        policy_hash_match = current_policy_hash == base_policy_hash

    normalized_trend_mode = _normalize_optional_text(trend_mode) or "gate_only"
    normalized_trend_reason = _normalize_optional_text(trend_reason) or "unknown"

    return {
        "mode": normalized_trend_mode,
        "reason": normalized_trend_reason,
        "required": _normalize_optional_bool(trend_required) is True,
        "executed": normalized_trend_mode == "gate_and_trend",
        "baseline_available": _normalize_optional_bool(baseline_available),
        "base_sha": _normalize_optional_text(base_sha),
        "policy_blob_current": _normalize_optional_text(current_policy_blob),
        "policy_blob_base": _normalize_optional_text(base_policy_blob),
        "policy_blob_match": _normalize_optional_bool(policy_blob_match),
        "policy_hash_current": current_policy_hash,
        "policy_hash_base": base_policy_hash,
        "policy_hash_match": policy_hash_match,
    }


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Attach skill-router trend metadata onto CI report JSON."
    )
    parser.add_argument(
        "--report",
        type=Path,
        default=Path("gateway/evals/data/skill_router_ci_report.json"),
        help="path to current skill-router report JSON",
    )
    parser.add_argument(
        "--base-report",
        type=Path,
        default=Path("gateway/evals/data/skill_router_ci_report.base.json"),
        help="path to baseline skill-router report JSON",
    )
    parser.add_argument("--trend-mode", default="gate_only", help="trend mode")
    parser.add_argument("--trend-reason", default="unknown", help="trend reason")
    parser.add_argument("--trend-required", default="false", help="whether trend check was required")
    parser.add_argument("--baseline-available", default="false", help="whether baseline report exists")
    parser.add_argument("--base-sha", default="", help="baseline commit SHA")
    parser.add_argument("--current-policy-blob", default="", help="current policy blob SHA")
    parser.add_argument("--base-policy-blob", default="", help="baseline policy blob SHA")
    parser.add_argument("--policy-blob-match", default="unknown", help="whether policy blob matched")
    parser.add_argument("--print-json", action="store_true", help="print trend metadata JSON payload")
    return parser


def main() -> None:
    parser = _build_parser()
    args = parser.parse_args()

    payload = _load_report(args.report)
    base_payload = _load_report(args.base_report)

    trend_meta = build_skill_router_trend_meta(
        current_report=payload,
        base_report=base_payload,
        trend_mode=args.trend_mode,
        trend_reason=args.trend_reason,
        trend_required=args.trend_required,
        baseline_available=args.baseline_available,
        base_sha=args.base_sha,
        current_policy_blob=args.current_policy_blob,
        base_policy_blob=args.base_policy_blob,
        policy_blob_match=args.policy_blob_match,
    )
    payload["trend_meta"] = trend_meta

    args.report.parent.mkdir(parents=True, exist_ok=True)
    with args.report.open("w", encoding="utf-8") as handle:
        json.dump(payload, handle, ensure_ascii=False, indent=2)
        handle.write("\n")

    if args.print_json:
        print(json.dumps({"skill_router_trend_meta": trend_meta}, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: gateway/src/evals/skill-router-trend-meta.ts
import { existsSync, readFileSync, writeFileSync, mkdirSync } from "node:fs";

const ZERO_SHA = "0000000000000000000000000000000000000000";

type JsonObject = Record<string, unknown>;

export interface SkillRouterTrendMetaInput {
  currentReport: JsonObject;
  baseReport: JsonObject;
  trendMode: unknown;
  trendReason: unknown;
  trendRequired: unknown;
  baselineAvailable: unknown;
  baseSha: unknown;
  currentPolicyBlob: unknown;
  basePolicyBlob: unknown;
  policyBlobMatch: unknown;
}

interface ParsedCliArgs {
  reportPath: string;
  baseReportPath: string;
  trendMode: string;
  trendReason: string;
  trendRequired: string;
  baselineAvailable: string;
  baseSha: string;
  currentPolicyBlob: string;
  basePolicyBlob: string;
  policyBlobMatch: string;
  printJson: boolean;
}

function readArgValue(argv: string[], index: number, flag: string): string {
  const value = argv[index + 1] ?? "";
  if (!value || value.startsWith("--")) {
    throw new Error(`missing value for ${flag}`);
  }
  return value;
}

function parseArgs(argv: string[]): ParsedCliArgs {
  const args: ParsedCliArgs = {
    reportPath: "gateway/evals/data/skill_router_ci_report.json",
    baseReportPath: "gateway/evals/data/skill_router_ci_report.base.json",
    trendMode: "gate_only",
    trendReason: "unknown",
    trendRequired: "false",
    baselineAvailable: "false",
    baseSha: "",
    currentPolicyBlob: "",
    basePolicyBlob: "",
    policyBlobMatch: "unknown",
    printJson: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    switch (token) {
      case "--report":
        args.reportPath = readArgValue(argv, index, "--report");
        index += 1;
        break;
      case "--base-report":
        args.baseReportPath = readArgValue(argv, index, "--base-report");
        index += 1;
        break;
      case "--trend-mode":
        args.trendMode = readArgValue(argv, index, "--trend-mode");
        index += 1;
        break;
      case "--trend-reason":
        args.trendReason = readArgValue(argv, index, "--trend-reason");
        index += 1;
        break;
      case "--trend-required":
        args.trendRequired = readArgValue(argv, index, "--trend-required");
        index += 1;
        break;
      case "--baseline-available":
        args.baselineAvailable = readArgValue(argv, index, "--baseline-available");
        index += 1;
        break;
      case "--base-sha":
        args.baseSha = readArgValue(argv, index, "--base-sha");
        index += 1;
        break;
      case "--current-policy-blob":
        args.currentPolicyBlob = readArgValue(argv, index, "--current-policy-blob");
        index += 1;
        break;
      case "--base-policy-blob":
        args.basePolicyBlob = readArgValue(argv, index, "--base-policy-blob");
        index += 1;
        break;
      case "--policy-blob-match":
        args.policyBlobMatch = readArgValue(argv, index, "--policy-blob-match");
        index += 1;
        break;
      case "--print-json":
        args.printJson = true;
        break;
      default:
        throw new Error(`unknown argument: ${token}`);
    }
  }
  return args;
}

function normalizeOptionalText(value: unknown): string | undefined {
  if (typeof value !== "string") {
    return undefined;
  }
  const normalized = value.trim();
  if (!normalized || normalized === ZERO_SHA) {
    return undefined;
  }
  return normalized;
}

function normalizeOptionalBool(value: unknown): boolean | undefined {
  if (typeof value === "boolean") {
    return value;
  }
  if (typeof value !== "string") {
    return undefined;
  }
  const normalized = value.trim().toLowerCase();
  if (normalized === "true") {
    return true;
  }
  if (normalized === "false") {
    return false;
  }
  return undefined;
}

function extractPolicyHash(report: JsonObject): string | undefined {
  const policyRaw = report.policy;
  if (typeof policyRaw === "object" && policyRaw !== null && !Array.isArray(policyRaw)) {
    const hash = normalizeOptionalText((policyRaw as JsonObject).hash);
    if (hash) {
      return hash;
    }
  }
  return normalizeOptionalText(report.policy_hash);
}

function asObject(value: unknown): JsonObject {
  if (typeof value === "object" && value !== null && !Array.isArray(value)) {
    return value as JsonObject;
  }
  return {};
}

export function loadReport(path: string): JsonObject {
  if (!existsSync(path)) {
    return {};
  }
  try {
    const raw = readFileSync(path, "utf8");
    return asObject(JSON.parse(raw));
  } catch {
    return {};
  }
}

function dirname(path: string): string {
  const normalized = path.replace(/[\\/]+$/, "");
  const slash = normalized.lastIndexOf("/");
  if (slash <= 0) {
    return ".";
  }
  return normalized.slice(0, slash);
}

export function saveReport(path: string, payload: JsonObject): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, `${JSON.stringify(payload, undefined, 2)}\n`, "utf8");
}

export function buildSkillRouterTrendMeta(input: SkillRouterTrendMetaInput): JsonObject {
  const currentPolicyHash = extractPolicyHash(input.currentReport);
  const basePolicyHash = extractPolicyHash(input.baseReport);
  let policyHashMatch: boolean | undefined;
  if (currentPolicyHash && basePolicyHash) {
    policyHashMatch = currentPolicyHash === basePolicyHash;
  }
  const normalizedTrendMode = normalizeOptionalText(input.trendMode) ?? "gate_only";
  const normalizedTrendReason = normalizeOptionalText(input.trendReason) ?? "unknown";
  return {
    mode: normalizedTrendMode,
    reason: normalizedTrendReason,
    required: normalizeOptionalBool(input.trendRequired) === true,
    executed: normalizedTrendMode === "gate_and_trend",
    baseline_available: normalizeOptionalBool(input.baselineAvailable),
    base_sha: normalizeOptionalText(input.baseSha) ?? null,
    policy_blob_current: normalizeOptionalText(input.currentPolicyBlob) ?? null,
    policy_blob_base: normalizeOptionalText(input.basePolicyBlob) ?? null,
    policy_blob_match: normalizeOptionalBool(input.policyBlobMatch),
    policy_hash_current: currentPolicyHash ?? null,
    policy_hash_base: basePolicyHash ?? null,
    policy_hash_match: policyHashMatch ?? null,
  };
}

export function runCli(argv: string[]): number {
  const args = parseArgs(argv);
  const report = loadReport(args.reportPath);
  const baseReport = loadReport(args.baseReportPath);
  const trendMeta = buildSkillRouterTrendMeta({
    currentReport: report,
    baseReport,
    trendMode: args.trendMode,
    trendReason: args.trendReason,
    trendRequired: args.trendRequired,
    baselineAvailable: args.baselineAvailable,
    baseSha: args.baseSha,
    currentPolicyBlob: args.currentPolicyBlob,
    basePolicyBlob: args.basePolicyBlob,
    policyBlobMatch: args.policyBlobMatch,
  });
  report.trend_meta = trendMeta;
  saveReport(args.reportPath, report);
  if (args.printJson) {
    process.stdout.write(`${JSON.stringify({ skill_router_trend_meta: trendMeta }, undefined, 0)}\n`);
  }
  return 0;
}

const entryScript = process.argv[1] ?? "";
const shouldRunCli = entryScript.includes("skill-router-trend-meta");

if (shouldRunCli) {
  try {
    process.exitCode = runCli(process.argv.slice(2));
  } catch (error) {
    process.stderr.write(`skill-router-trend-meta fatal: ${String(error)}\n`);
    process.exitCode = 1;
  }
}
